File: csharp/955.delete-columns-to-make-sorted-ii.cs
/*
 * @lc app=leetcode id=955 lang=csharp
 *
 * [955] Delete Columns to Make Sorted II
 */

using System.Collections.Generic;
using System.Linq;

// @lc code=start
public class Solution {
    public int MinDeletionSize(string[] A) {
        int n = A.Length, m = A[0].Length, ret = 0;
        bool[] stop = new bool[n];
        for (int j = 0; j < m; ++j) {
            var chars = A.Select(s => s[j]);
            if (chars
                .Zip(chars.Skip(1), (char curr, char next) => curr > next)
                .Zip(stop, (bool b1, bool b2) => b1 && !b2)
                .Any(b => b)) {
                ++ret;
            } else {
                for (int i = 0; i < n - 1; ++i) {
                    if (A[i][j] < A[i + 1][j]) {
                        stop[i] = true;
                    }
                }
            }
        }
        return ret;
    }
}
// @lc code=end

File: java/955.delete-columns-to-make-sorted-ii.java
/*
 * @lc app=leetcode id=955 lang=java
 *
 * [955] Delete Columns to Make Sorted II
 */

// @lc code=start
class Solution {
    public int minDeletionSize(String[] strs) {
        int n = strs.length, m = strs[0].length(), ret = 0;
        boolean[] sorted = new boolean[n - 1];
        for (int j = 0; j < m; ++j) {
            int i = 0;
            for (; i < n - 1; ++i) {
                if (sorted[i]) {
                    continue;
                }
                if (strs[i].charAt(j) > strs[i + 1].charAt(j)) {
                    ++ret;
                    break;
                }
            }
            if (i < n - 1) {
                continue;
            }
            for (i = 0; i < n - 1; ++i) {
                sorted[i] |= strs[i].charAt(j) < strs[i + 1].charAt(j);
            }
        }
        return ret;
    }
}
// @lc code=end

